src/System.Drawing.Design.Primitives/src/System/Drawing/Design/ToolboxItem.cs

```diff
@@ -0,0 +1,1100 @@
+            }
+
+            //first search for an exact match -- we prefer this over a partial match.
+            foreach (AssemblyName name in type.Assembly.GetReferencedAssemblies())
+            {
+                if (name.FullName == policiedAssemblyName.FullName)
+                    return name;
+            }
+
+            //next search for a partial match -- we just compare the Name portions (ignore version and publickey)
+            foreach (AssemblyName name in type.Assembly.GetReferencedAssemblies())
+            {
+                if (name.Name == policiedAssemblyName.Name)
+                    return name;
+            }
+
+            //finally, the most expensive -- its possible that retargeting policy is on an assembly whose name changes
+            // an example of this is the device System.Windows.Forms.Datagrid.dll
+            // in this case, we need to try to load each device assemblyname through policy to see if it results
+            // in assemblyname.
+            foreach (AssemblyName name in type.Assembly.GetReferencedAssemblies())
+            {
+                Assembly a = null;
+
+                try
+                {
+                    a = Assembly.Load(name);
+                    if (a != null && a.FullName == policiedAssemblyName.FullName)
+                        return name;
+                }
+                catch
+                {
+                    //ignore all exceptions and just fall through if it fails (it shouldn't, but who knows).
+                }
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Initializes a toolbox item with a given type.  A locked toolbox item cannot be initialized.
+        /// </summary>
+        public virtual void Initialize(Type type)
+        {
+            CheckUnlocked();
+
+            if (type != null)
+            {
+                TypeName = type.FullName;
+                AssemblyName assemblyName = type.Assembly.GetName(true);
+                if (type.Assembly.GlobalAssemblyCache)
+                {
+                    assemblyName.CodeBase = null;
+                }
+
+                Dictionary<string, AssemblyName> parents = new Dictionary<string, AssemblyName>();
+                Type parentType = type;
+                while (parentType != null)
+                {
+                    AssemblyName policiedname = parentType.Assembly.GetName(true);
+
+                    AssemblyName aname = GetNonRetargetedAssemblyName(type, policiedname);
+
+                    if (aname != null && !parents.ContainsKey(aname.FullName))
+                    {
+                        parents[aname.FullName] = aname;
+                    }
+                    parentType = parentType.BaseType;
+                }
+
+                AssemblyName[] parentAssemblies = new AssemblyName[parents.Count];
+                int i = 0;
+                foreach (AssemblyName an in parents.Values)
```

Comment:
can this be done with .ToArray?

Reply:
this pattern is  used in multiple places. If possible, i want to leave them as is as part of this port..

Reply:
I'm fine with it since this is a port, but just wanted to call it out, any cleanup we can do when porting is nice.